### PR TITLE
Restcomm 1861 sl return effective profile

### PR DESCRIPTION
**Pull request by maria-farooq**

### CODE REVIEW

File: restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/api/NumberSelectorService.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     *
+     * @param result whether the call should be rejected depending on results
+     * found
+     * @param srcOrg
+     * @param destOrg
+     * @return
+     */
+    boolean isFailedCall(NumberSelectionResult result, Sid srcOrg, Sid destOrg);
```
Comment (jaimecasero) [inline restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/api/NumberSelectorService.java]: stricly speaking, we should make NumberSelectionResult  part of api pck , so clients doesnt need to import impl pck

File: restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/api/ProfileService.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import com.sun.jersey.core.header.LinkHeader;
+import javax.ws.rs.core.UriInfo;
+import org.restcomm.connect.dao.entities.Profile;
+
+public interface ProfileService {
+    /**
+     * @param accountSid
+     * @return  will return associated profile of provided account sid
+     */
+    Profile retrieveEffectiveProfileByAccountSid(String accountSid);
```
Comment (jaimecasero) [inline restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/api/ProfileService.java]: i would try to consolidate using Sid class. using string for Sid reveals Sid details that may not be applicable later

File: restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/profile/ProfileServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    /**
+     * @param targetSid
+     * @param info
+     * @param resource
+     * @return
+     */
+    @Override
+    public LinkHeader composeProfileLink(String targetSid, UriInfo info, Class resource) {
```
Comment (jaimecasero) [inline restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/profile/ProfileServiceImpl.java]: not sure whether serviec layer should be aware of REST endpoint stuff, with this we need jaxrs dependency in core lib...

File: restcomm/restcomm.application/src/main/java/org/restcomm/connect/application/Bootstrapper.java [hunk 4]
```diff
@@ -401,8 +402,8 @@ public void servletInitialized(SipServletContextEvent event) {
             IdentityContext identityContext = new IdentityContext(xml);
             context.setAttribute(IdentityContext.class.getName(), identityContext);
 
-            //init NumberSelectorService
-            context.setAttribute(NumberSelectorService.class.getName(), new NumberSelectorService(storage.getIncomingPhoneNumbersDao()));
+            // Initialize CoreServices
+            RestcommConnectServiceProvider.getInstance().startServices(storage);
```
Comment (jaimecasero) [inline restcomm/restcomm.application/src/main/java/org/restcomm/connect/application/Bootstrapper.java]: not sure if we should still publish services in servlet context for consistency purposes...

File: restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/RestcommConnectServiceProvider.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        }
+        return instance;
+    }
+
+    /**
+     * @param daoManager
+     */
+    public void startServices(DaoManager daoManager) {
+        // core services initialization
+        this.numberSelector = new NumberSelectorServiceImpl(daoManager.getIncomingPhoneNumbersDao());
```
Comment (jaimecasero) [inline restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/RestcommConnectServiceProvider.java]: should we publish services in servletContext as usual?

Comment (maria-farooq) [reply restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/RestcommConnectServiceProvider.java]: purpose to have it in servletContext is mainly access it from project, while in current case we can easily access them by an instance of this class

File: restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/api/NumberSelectorService.java (hunk 1 above)
Comment (maria-farooq) [reply restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/api/NumberSelectorService.java]: right, will move it

Comment (maria-farooq) [reply restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/api/NumberSelectorService.java]: i will move it in `org.restcomm.connect.core.service.number.api`

File: restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/api/ProfileService.java (hunk 2 above)
Comment (maria-farooq) [reply restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/api/ProfileService.java]: agreed

File: restcomm/restcomm.application/src/main/java/org/restcomm/connect/application/Bootstrapper.java (hunk 4 above)
Comment (maria-farooq) [reply restcomm/restcomm.application/src/main/java/org/restcomm/connect/application/Bootstrapper.java]: i would not consider it a point of consistency

File: restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/RestcommConnectServiceProvider.java (hunk 5 above)
Comment (jaimecasero) [reply restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/RestcommConnectServiceProvider.java]: static singleton are bad for mocking/testing. currently mockito cant mock static getInstance invocations, so this means we will not able to mock service as a collaborator in any potential client. I think this is pretty bad result...

**Comment by gvagenas** [comment]
@maria-farooq Patch looks good to me, you can proceed with the merge.

**Comment by gvagenas** [comment]
@maria-farooq I merged with master to get latest fixes I pushed (related to RESTCOMM-1580)

**Pull request merged**